### DISCUSSION TITLE
Observer-task running continuesly - best practice?

### DISCUSSION THREAD

**Discussion by thp44** [Miksus/rocketry#138] [Q&A] [answered]
Hi,
I have a observer-task I want to always run with my other time based tasks. What would be the best practice for a continuously running "task"/observer in rocketry?
Can't seem to find it in the documentation.

**Comment by Miksus** [comment, marked as answer]
Hi!

Hmm, it indeed seems I have forgotten to mention about the ``permanent_task``:

```python
@app.task(permanent_task=True, on_startup=True, execution="async")
async def do_constantly():
    while True:
        ...
        await asyncio.sleep(0)
```

There are a couple of reasons to use this. Without it, the task risks being timeout. If you put infinite as the timeout, there is a problem in the shutdown: the scheduler waits for the task to finish by default (which will never happen). This attribute signals to the scheduler that the task should never timeout but it should not wait it to finish when the scheduler shuts down.

You can use any execution with the task except ``main`` as then the task blocks everything. Also, note that I used the ``on_startup=True``, this by default makes the task be started in the startup cycle (no need to put a start condition).

I possibly rename the ``permanent_task`` to ``permanent`` in the future (as the word *task* is pointless there) but I'll introduce a warning when that happens.

Did I answer to the question?

**Reply by thp44** [reply]
Yup! Thank you very much for quick and detailed answer. Can't wish for more. 💯 👍

**Reply by Miksus** [reply]
No problem! Thanks for the question by the way. I'll make an issue to improve the documentation about the permanent tasks (as there is indeed no mention of those).
